### PR TITLE
Persistent login between page refreshes/re-routing #641

### LINKED ISSUE #641: Login token information ignored when page is refreshed
**Description**:
If a use logs in (via icat authenticator at least) and then refreshes the page - their token is still valid but certain parts of the app do not use the information in the token payload. e.g. their username is not displayed in the user profile, and wrt #591 they "lose" their admin functionality.

**Acceptance criteria**:
- [ ] Login token payload information is retained in scigateway after a page refresh

### PR DESCRIPTION
**Pull request by sam-glendenning**
## Description
Keeps the user logged in after refreshing the web page or navigating to a different address. On page refresh, when auth provider is constructed in code, a check is made for an existing token. If one is found, the user details are parsed from it and stored in code. Otherwise, user is assumed not to be logged in.

No additional data is stored locally, relies on the token being secure and signed. Any changes to the token made by the user invalidate the session and log the user out. This prevents the user from modifying their username or admin status.

Also amending test data in auth provider test code to use a valid JWT as test data instead of the string 'token'. This is to allow for testing to be carried out to confirm user login. The code is unable to parse user details from a token string equalling 'token' so a JWT with test data is used instead.

## Testing instructions
Test with JWT authenticator (username = 'username') and ICAT authenticator (username = 'root'). 

- [x] Review code
- [x] Check Actions build
- [x] After login, validate user that is logged in and their functionality. Then refresh the page or navigate to a different one to check the same functionality exists
- [x] If logged in as admin user, check navigation to /admin works after page refresh
- [x] Check token refreshing and validation is happening with requests to `/checkToken` and `/refresh` for JWT and `/verify` and `/refresh` for ICAT
- [x] Check that overwriting `scigateway:token` value in localStorage with a user-input value logs the user out on a page refresh
- [x] Review changes to test coverage

## Agile board tracking
closes #641

### CODE REVIEW

**Review by louise-davies — CHANGES_REQUESTED** [review]
Looks good - just a few comments re: typing declarations & removing some commented out code

File: src/authentication/baseAuthProvider.tsx [hunk 1]
```diff
@@ -14,11 +15,35 @@ export default abstract class BaseAuthProvider implements AuthProvider {
...
+      try {
+        const tokenString: string = parseJwt(this.token);
+        if (tokenString) {
+          const tokenObject = JSON.parse(tokenString);
+          const user: User = new UserInfo(tokenObject.username);
+          user.isAdmin = tokenObject.userIsAdmin;
+          return user;
+        }
+      } catch (TypeError) {
+        // not a valid JWT, token has likely been tampered with in some way (or we are running tests)
```
[the remark this answers is not in the document]
Comment (sam-glendenning) [reply src/authentication/baseAuthProvider.tsx:40]: In the console? Or somewhere else?

File: cypress/integration/login.spec.ts [hunk 2]
```diff
@@ -109,6 +109,57 @@ describe('Login', () => {
...
+    cy.window().then(
+      (window) => {
+        expect(window.localStorage.getItem('scigateway:token')).not.be.null;
+        expect(storedToken).to.equal(window.localStorage.getItem('scigateway:token'));
+      });
+    cy.window().then((window) =>
+      expect(storedToken).to.equal(
+        window.localStorage.getItem('scigateway:token')
+      )
+    );
```
[the remark this answers is not in the document]
Comment (sam-glendenning) [reply cypress/integration/login.spec.ts:150]: I'd forgotten I'd added that in. I'll remove the later check, thanks for spotting that.

**Review by louise-davies — APPROVED** [review]
Looks good 👍

**Pull request merged**